src/main/java/net/datafaker/service/WeightedRandomSelector.java

```diff
@@ -0,0 +1,126 @@
+    private static final String WEIGHT_KEY = "weight";
+    private static final String VALUE_KEY = "value";
+    private final Random random;
+
+    public WeightedRandomSelector(Random random) {
+        this.random = random != null ? random : new Random();
+    }
+
+    public Random getRandom() {
+        return random;
+    }
+
+    /**
+     * Returns a weighted random element from the given list, where each element is represented as a Map
+     * containing a weight and the corresponding value.
+     * <p>
+     * @param items A list of maps, where each map contains:
+     *              - weight: A Double representing the weight of the element, influencing its selection probability.
+     *              - value: The actual element of type T to be randomly selected based on its weight.
+     * @param <T> The type of the element to be selected from the list. The value associated with the weight can be of any type.
+     * @return A randomly selected element based on its weight.
+     * @throws IllegalArgumentException if:
+     *                                  - the list is null or empty,
+     *                                  - any item in the list is null or empty,
+     *                                  - the item does not contain 'weight' or 'value' keys,
+     *                                  - any weight is null, non-positive, NaN or infinite,
+     *                                  - any values in the list are not unique or null,
+     *                                  - the sum of weights exceeds Double.MAX_VALUE.
+     */
+    public <T> T select(List<Map<String, Object>> items) {
+        validateItemsList(items);
+
+        Object[] values = new Object[items.size()];
+        double[] cumulativeWeights = preprocessItems(items, values);
+
+        double randomValue = random.nextDouble() * cumulativeWeights[cumulativeWeights.length - 1];
+        return selectWeightedElement(randomValue, cumulativeWeights, values);
+    }
+
+    private void validateItemsList(List<Map<String, Object>> items) {
+        if (items == null || items.isEmpty()) {
+            throw new IllegalArgumentException("weightedArrayElement expects a non-empty list");
+        }
+
+        Set<Object> uniqueValues = new HashSet<>();
+
+        for (var item : items) {
+            validateItem(item);
+            assertUniqueValues(item, uniqueValues);
+        }
+    }
+
+    private void assertUniqueValues(Map<String, Object> item, Set<Object> values) {
+        Object value = item.get(VALUE_KEY);
+        if (!values.add(value)) {
+            throw new IllegalArgumentException("Duplicate value found: " + value + ". Values must be unique.");
+        }
+    }
+
+    private void validateItem(Map<String, Object> item) {
+        if (item == null || item.isEmpty()) {
+            throw new IllegalArgumentException("Item cannot be null or empty");
+        }
+        if (!item.containsKey(WEIGHT_KEY) || !item.containsKey(VALUE_KEY)) {
```

Comment:
are we checking separately existing of keys and non null values only for the sake of different error messages>

Reply:
I split the condition into separate checks with distinct error messages.